### PR TITLE
feature(schemas): coded values mapping

### PR DESCRIPTION
**Pull request by vitaligi**
adds the ability to map tag values to a coded value based on a mapping in remote repo (redis)

| Question                | Answer                                                                          |
| ---------------- | -------------------------------------------------------------------------- |
| Bug fix         | ✖                                                                        |
| New feature     | ✔                                                                        |
| Breaking change | ✖                                                                        |
| Deprecations    | ✖                                                                        |
| Documentation   | ✖                                                                        |
| Tests added     | ✔                                                                        |
| Chore            | ✖                                                                       |

### CODE REVIEW

File: src/common/db.ts [hunk 1]
```diff
@@ -1,18 +1,31 @@
...
+      }
+      redis = new Redis(url.href);
+    }
     await redis.connect();
     return redis;
   } catch (e) {
+    if (e instanceof TypeError) {
+      throw e;
+    }
     throw new Error('Redis connection failed');
```
Comment (CptSchnitz) [inline src/common/db.ts:29]: i dont think its necessary to catch the error here, please check and remove if so.

File: src/schema/DAL/domainFieldsRepository.ts [hunk 2]
```diff
@@ -1,6 +1,6 @@
 export const IDOMAIN_FIELDS_REPO_SYMBOL = Symbol('DOMAINFIELDSREPO');
 
 export interface IDomainFieldsRepository {
-  getFields: (fields: string[]) => Promise<string[]>;
+  getFields: ((fields: string[], key: string | undefined) => Promise<string[]>) | ((fields: string[]) => Promise<string[]>);
```
Comment (CptSchnitz) [inline src/schema/DAL/domainFieldsRepository.ts:4]: if key is constant, dont ask for it on every function call.

File: src/schema/DAL/redisManager.ts [hunk 3]
```diff
@@ -14,10 +14,12 @@ export class RedisManager implements IDomainFieldsRepository {
...
   }
 
-  public async getFields(fields: string[]): Promise<string[]> {
+  public async getFields(fields: string[], key: string | undefined): Promise<string[]>;
+  public async getFields(fields: string[]): Promise<string[]>;
+  public async getFields(fields: string[], key?: string): Promise<string[]> {
     let res: (string | null)[];
     try {
-      res = await this.redis.mget(fields);
+      res = key !== undefined ? await this.redis.hmget(key, fields) : await this.redis.mget(fields);
```
Comment (CptSchnitz) [inline src/schema/DAL/redisManager.ts:22]: change it to multiline if please.

File: src/schema/models/schemaManager.ts [hunk 4]
```diff
@@ -125,32 +130,36 @@ export class SchemaManager {
...
-    const redisRes = await this.domainFieldsRepo.getFields(redisKeysArr);
+    const fieldsCodedValuesRes = await this.domainFieldsRepo.getFields(domainFieldsKeys, hashKey);
 
     //for each domain field create new domain field tag with the correct value
-    redisRes.forEach((key, index) => {
+    fieldsCodedValuesRes.forEach((codedValue, index) => {
       domainFieldsTags = {
         ...domainFieldsTags,
-        [redisKeysArr[index].split(':')[0] + '_DOMAIN']: key,
+        [domainFieldsKeys[index].split(':')[0]]: codedValue,
```
Comment (CptSchnitz) [inline src/schema/models/schemaManager.ts:141]: revert back to adding _domain.

File: src/schema/models/schemaManager.ts [hunk 5]
```diff
@@ -125,32 +130,36 @@ export class SchemaManager {
...
-    redisRes.forEach((key) => {
-      let explode = JSON.parse(key) as Record<string, string | number | null>;
-      explode = Object.entries(explode).reduce((acc, [key, value]) => {
-        return { ...acc, [key]: value };
-      }, {});
-      explodeFieldsTags = { ...explodeFieldsTags, ...explode };
+    explodeRes.forEach((json, index) => {
+      try {
+        let explode = JSON.parse(json) as Record<string, string | number | null>;
+        explode = Object.entries(explode).reduce((acc, [key, value]) => {
```
Comment (CptSchnitz) [inline src/schema/models/schemaManager.ts:156]: only from line 156 it should be called exploded

**Review by CptSchnitz — CHANGES_REQUESTED** [review]

File: src/schema/models/schemaManager.ts (hunk 4 above)
Comment (melancholiai) [reply src/schema/models/schemaManager.ts]: done.

**Comment by vitaligi** [comment]
closes #101

**Pull request merged**